deepinv/distrib/distrib_framework.py

```diff
@@ -0,0 +1,804 @@
+        :param Callable factory: factory function that creates linear physics operators.
+        :param None, dict shared: shared data dictionary passed to factory function.
+        :param str reduction: reduction mode for distributed operations. Options are `'sum'` and `'mean'`.
+        :param None, torch.dtype dtype: data type for operations.
+        :param str gather_strategy: strategy for gathering distributed results.
+        """
+        LinearPhysics.__init__(self, A=lambda x: x, A_adjoint=lambda y: y, **kwargs)
+        self.reduction_mode = reduction
+        super().__init__(ctx, num_ops, factory, shared=shared, dtype=dtype, 
+                        gather_strategy=gather_strategy, **kwargs)
+
+        for p in self.local_physics:
+            if not isinstance(p, LinearPhysics):
+                raise ValueError("factory must return LinearPhysics instances.")
+
+    # ---- local (fast) ----
+    def A_adjoint_local(self, y_local: list[torch.Tensor], **kwargs) -> torch.Tensor:
+        if len(y_local) == 0:
+            # Return zeros with proper shape for empty local set
+            return torch.zeros((), device=self.ctx.device, dtype=self.dtype)
+        contribs = [
+            p.A_adjoint(y_i, **kwargs)
+            for p, y_i in zip(self.local_physics, y_local, strict=False)
+        ]
+        return torch.stack(contribs, dim=0).sum(0)
+
+    def A_vjp_local(
+        self, x: torch.Tensor, v_local: list[torch.Tensor], **kwargs
+    ) -> torch.Tensor:
+        if len(v_local) == 0:
+            # Return zeros with proper shape for empty local set
+            return torch.zeros_like(x)
+        contribs = [
+            p.A_vjp(x, v_i, **kwargs)
+            for p, v_i in zip(self.local_physics, v_local, strict=False)
+        ]
+        return torch.stack(contribs, dim=0).sum(0)
+
+    # ---- global (compat) ----
+    def _reduce_global(self, x_like: torch.Tensor) -> torch.Tensor:
+        if not torch.is_tensor(x_like):
+            # handle 0.0 placeholder for empty local set
+            x_like = torch.zeros((), device=self.ctx.device, dtype=self.dtype)
+            x_like = x_like.expand(())  # scalar
+
+        # Ensure all ranks participate in collective even with empty data
+        if self.ctx.is_dist:
+            # For ranks with empty local sets, x_like should be zeros
+            self.ctx.all_reduce_(x_like, op="sum")
+
+        if self.reduction_mode == "mean":
+            x_like = x_like / float(self.num_ops)
+        return x_like
+
+    def A_adjoint(self, y: TensorList, **kwargs) -> torch.Tensor:
+        y_local = [y[i] for i in self.local_idx]
+        local = self.A_adjoint_local(y_local, **kwargs)
+        return self._reduce_global(local)
+
+    def A_vjp(self, x: torch.Tensor, v: TensorList, **kwargs) -> torch.Tensor:
+        v_local = [v[i] for i in self.local_idx]
+        local = self.A_vjp_local(x, v_local, **kwargs)
+        return self._reduce_global(local)
+
+
```

Comment:
Anticipating a potential call to `A_adjoint_A` (eg, in `least_squares` solvers), I would provide an efficient implementation instead of relying on `super().A_adjoint_A`, which naively involves 2 gather/reduce operations, when, if I'm not mistaken, only one should be sufficient in practice ? 

Could something like this work ? 
```suggestion

    def A_adjoint_A_local(self, x: torch.Tensor, **kwargs) -> torch.Tensor:
        if len(self.local_physics) == 0:
            return torch.zeros_like(x)
        contribs = [
            p.A_adjoint_A(x, **kwargs) for p in self.local_physics
        ]
        return torch.stack(contribs, dim=0).sum(0)

    def A_adjoint_A(self, x: torch.Tensor, **kwargs) -> torch.Tensor:
        local = self.A_adjoint_A_local(x, **kwargs)
        return self._reduce_global(local)
```